File: src/main/java/name/seguri/java/leetcode/_49/Solution.java
package name.seguri.java.leetcode._49;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
  public List<List<String>> groupAnagrams(String[] strs) {
    Map<String, List<String>> occurrences = new HashMap<>();

    for (String s : strs) {
      var sortedString = sort(s);
      occurrences.computeIfAbsent(sortedString, k -> new ArrayList<>()).add(s);
    }

    return new ArrayList<>(occurrences.values());
  }

  private String sort(String s) {
    var chars = s.toCharArray();
    Arrays.sort(chars);
    return new String(chars);
  }
}

File: src/main/kotlin/name/seguri/kotlin/leetcode/_49/Solution.kt
package name.seguri.kotlin.leetcode._49

class Solution {
  fun groupAnagrams(strs: Array<String>): List<List<String>> {
    val occurrences = mutableMapOf<String, MutableList<String>>()

    for (s in strs) {
      val sortedString = s.toCharArray().sorted().joinToString("")
      occurrences.getOrPut(sortedString) { mutableListOf() }.add(s)
    }

    return occurrences.values.toList()
  }
}
